### PR TITLE
[WIP] Refactoring replication

### PR DESCRIPTION
**Pull request by kosssi**
General:
- [ ] remove conflict for pouchdb document
- [ ] add comments
- [ ] add tests

### CODE REVIEW

File: www-src/app/change/change_manager.coffee [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+###*
+ * ChangeManager allows to launch the good manager with specific state
+ *
+ * @class ChangeManager
+###
+module.exports = class ChangeManager
+
+    ###*
+     * Create a FilterManager.
```
Comment (nono) [inline www-src/app/change/change_manager.coffee]: comments

File: www-src/app/change/change_manager.coffee [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+log = require('../lib/persistent_log')
+    prefix: "Change Manager"
+    date: true
+
+###*
+ * ChangeManager allows to launch the good manager with specific state
+ *
+ * @class ChangeManager
+###
+module.exports = class ChangeManager
```
Comment (nono) [inline www-src/app/change/change_manager.coffee]: https://blog.codinghorror.com/i-shall-call-it-somethingmanager/

File: www-src/app/replicator/replicator_config.coffee [hunk 3]
```diff
@@ -83,5 +101,20 @@ module.exports = class ReplicatorConfig extends Backbone.Model
...
     hasPermissions: ->
-        _.isEqual @get('devicePermissions'), \
-            @serializePermissions(@replicator.permissions)
+        return false unless @permissions
+        _.isEqual @get('devicePermissions'), @serializePermissions(@permissions)
+
+    getFilterManager: ->
+        unless @filterManager
+            @filterManager = new FilterManager @getCozyUrl(), @get('auth'), \
+                    @get("deviceName")
```
Comment (nono) [inline www-src/app/replicator/replicator_config.coffee]: You can simplify these 3 lines with:

```coffee
@filterManager ?= new FilterManager ...
```

File: www-src/app/change/change_manager.coffee (hunk 1 above)
Comment (kosssi) [inline www-src/app/change/change_manager.coffee]: :smile: beautiful picture :+1:

File: www-src/app/change/change_manager.coffee (hunk 2 above)
Comment (frankrousseau) [inline www-src/app/change/change_manager.coffee]: Interesting. I think I do the same with word helper. When I don't know how to name a bunch of functions.

File: www-src/app/change/change_manager.coffee (hunk 1 above)
Comment (frankrousseau) [inline www-src/app/change/change_manager.coffee]: What do these cookies mean?

Comment (nono) [inline www-src/app/change/change_manager.coffee]: The cookies are in a box labeled "Basmati Rice". It's a metaphor to say that the comment and the code don't match. The comment speaks about `FilterManager` but, in the code, it's a `ChangeManager`.

Comment (frankrousseau) [inline www-src/app/change/change_manager.coffee]: I dare to write "LOL".

**Comment by kosssi** [comment]
I update the pull-request.

 * ChangeManager -> ChangeDispatcher
 * FileManager -> ChangeFileHandler
 * ReplicatorManager -> ReplicationLauncher

File: www-src/app/change/change_dispatcher.coffee [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    prefix: "ChangeDispatcher"
+    date: true
+
+###*
+ * ChangeDispatcher allows to launch the good manager with specific state
+ *
+ * @class ChangeDispatcher
+###
+module.exports = class ChangeDispatcher
+
```
Comment (jacquarg) [inline www-src/app/change/change_dispatcher.coffee]: We could simplify a lot this file using a dict :
```
@changeHadlers = 
    file: new ChangeFileHandler config
    event: new ChangeEventHandler
   ....

```
Dispatch would be two lines, and isDispatched one ; )

**Pull request closed without merging**